### PR TITLE
Datadog filter channel

### PR DESCRIPTION
**Pull request by Paul-Hess**

### CODE REVIEW

**Comment by Paul-Hess** [comment]
This will probably be the easiest to review, but it contains commits from 3 other PRs, may want to hold off from going through it until those are approved and not subject to change

File: src/main/java/com/flightstats/hub/metrics/StatsDFilter.java [hunk 1]
```diff
@@ -1,18 +1,19 @@
...
 import com.timgroup.statsd.NoOpStatsDClient;
 import com.timgroup.statsd.NonBlockingStatsDClient;
 import com.timgroup.statsd.StatsDClient;
 
 import java.util.Arrays;
 import java.util.Collections;
-import java.util.HashMap;
 import java.util.List;
-import java.util.Map;
+import java.util.Optional;
```
Comment (manjula-chikkanagappa) [inline src/main/java/com/flightstats/hub/metrics/StatsDFilter.java]: 👍

File: src/test/java/com/flightstats/hub/webhook/WebhookTest.java [hunk 2]
```diff
@@ -141,4 +147,12 @@ public void testIsTagPrototype() throws Exception {
         assertTrue(twh.isTagPrototype());
     }
 
+    @Test
+    public void testSecondaryMetricsReporting() {
+        Webhook withDefaults = this.webhook.withDefaults();
+        assertFalse(withDefaults.isSecondaryMetricsReporting());
+        String json = "{ \"secondaryMetricsReporting\": \"true\" }";
```
Comment (manjula-chikkanagappa) [inline src/test/java/com/flightstats/hub/webhook/WebhookTest.java]: is true a string or boolean? if boolean, quotes not required.

File: src/main/java/com/flightstats/hub/model/ChannelConfig.java [hunk 3]
```diff
@@ -46,10 +50,23 @@
...
+                          boolean keepForever,
+                          String description,
+                          Set<String> tags,
+                          String replicationSource,
+                          String storage,
+                          boolean protect,
+                          DateTime mutableTime,
+                          boolean allowZeroBytes,
+                          String displayName,
+                          String secondaryMetricsReporting) {
```
[the remark this answers is not in the document]
Comment (lkemmerer) [reply src/main/java/com/flightstats/hub/model/ChannelConfig.java]: I agree. This looks like one of the classes that got de-lombokified a while back. Assuming it doesn't have too much logic in it, I'd definitely convert it back.

[the remark this answers is not in the document]
Comment (lkemmerer) [reply src/main/java/com/flightstats/hub/model/ChannelConfig.java]: ...and if it does have too much logic in it, maybe that could be moved over to a different class if it's easy to extricate.

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java [hunk 4]
```diff
@@ -4,29 +4,25 @@
...
-import org.slf4j.LoggerFactory;
 
-import java.io.IOException;
 import java.util.Arrays;
 import java.util.List;
 
 import static org.junit.Assert.*;
 
 public class ChannelConfigTest {
-    private static final Logger logger = LoggerFactory.getLogger(ChannelConfigTest.class);
```
Comment (lkemmerer) [inline src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: On a tangent, I am not 100% sure why we use a logger factory all over the place. Seems like there'd be some sort of annotation we could use.

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java [hunk 5]
```diff
@@ -137,13 +133,13 @@ public void testCopy() throws IOException {
                 .mutableTime(TimeUtil.now())
                 .allowZeroBytes(false)
                 .build();
-        assertTrue(config.equals(config.toBuilder().build()));
 
-        assertTrue(config.equals(ChannelConfig.createFromJson(config.toJson())));
+        assertEquals(config, config.toBuilder().build());
```
Comment (lkemmerer) [inline src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: ...I don't know what `.toBuilder()` does and I'm a little bit afraid to look. I know you didn't introduce it, so don't feel obliged to answer if you don't know.

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java [hunk 6]
```diff
@@ -189,4 +185,34 @@ public void testZeroBytes() {
         assertFalse(testZeroBytes.isAllowZeroBytes());
     }
 
+    @Test
+    public void testSecondaryMetricsReporting_defaultEmpty() {
+        ChannelConfig testSecondaryMetricsNull = ChannelConfig
+                .builder()
+                .name("testSecondaryMetrics0")
+                .build();
+        assertEquals("", testSecondaryMetricsNull.getSecondaryMetricsReporting());
```
Comment (lkemmerer) [inline src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: If you change this to a lombok object, you'll either need to add a default value on this or write your own getter. I guess either one isn't too bad...

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 5 above)
Comment (chriskessel) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: toBuilder() is full of awesomeness. It creates an exact close of the original object, but returns a builder. Great for  "I need a clone, but with this one field changed" situations.

That said...no idea why you'd to a toBuilder() and then immediately .build().

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 6 above)
Comment (lkemmerer) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: On the bright side, if you just do a default value, you can probably get rid of the next test down.

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 4 above)
Comment (chriskessel) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: ```@Slf4j```

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 5 above)
Comment (lkemmerer) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: Ah, neat! I usually use the lombok annotation that lets you do `withAttribute()` chaining to accomplish something similar. ...As you pointed out, I agree that the main concern was the immediate `build()` that happens afterwards.

**Comment by lkemmerer** [comment]
Thanks for cleaning up those test assertions. :)

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 4 above)
Comment (lkemmerer) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: Ooh. I was afraid the reason we were doing this was that we weren't using that library. I'm absolutely going to start converting these. Thanks!

Comment (chriskessel) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: I think it should work as there's some slf4j bridge to most loggers that seems to magically happen (presumably somehow via lombok or it's dependencies). You'd have to verify it though.

Comment (lkemmerer) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: There are a couple annotations in the codebase already (I think Zander started adding them before he left). I'll double check that they're working.

Comment (lkemmerer) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: Yep, they're working just fine.
As a side note, I'm going to start a confluence doc with common cleanup things we should try to do in the codebase as we touch stuff.

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 5 above)
Comment (Paul-Hess) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: I am guessing the purpose of this was to test that the (all hand-rolled) hash and equals and builder and toBuilder worked in unison to create an exact copy. Which IIRC is implicitly tested on the next line as createFromJson uses that method under the hood.

File: src/test/java/com/flightstats/hub/webhook/WebhookTest.java (hunk 2 above)
Comment (Paul-Hess) [reply src/test/java/com/flightstats/hub/webhook/WebhookTest.java]: true :) pun intended the quotes are not necessary

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 4 above)
Comment (Paul-Hess) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: 👍

File: src/test/java/com/flightstats/hub/model/ChannelConfigTest.java (hunk 5 above)
Comment (lkemmerer) [reply src/test/java/com/flightstats/hub/model/ChannelConfigTest.java]: Ah, I guess it'd make sense to leave it there at least temporarily to make sure re-lombokifying things works out if you decide to make that change.

File: src/main/java/com/flightstats/hub/model/ChannelConfig.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (Paul-Hess) [reply src/main/java/com/flightstats/hub/model/ChannelConfig.java]: carving out this as a new PR against this one here https://github.com/flightstats/hub/tree/refactor-channel-config

**Pull request merged**